pkg/tagset/factory.go

```diff
@@ -0,0 +1,105 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+// The Factory type is responsible for creating new Tags instances. Its
+// interface is simple, but provides an opportunity for optimization and
+// deduplication.
+//
+// A single default factory is provided for use throughout the agent, with
+// package-level functions deferring to that factory. Additional, specific
+// factories may be created for specific purposes. The default factory is
+// thread-safe, but it may be advantageous to build non-thread-safe factories
+// for specific circumstances. Tags instances created by different factories
+// can be used interchangeably and are entirely thread-safe.
+//
+// Tags instances returned from different factories may be used
+// interchangeably. The only disadvantage of using multiple factories is a
+// reduced cache rate due to not sharing caches between those factories.
+//
+// This interface contains un-exported methods, so it cannot be implemented outside
+// of this package.
+type Factory interface {
+	// Tags constructors
+
+	// NewTags creates a new *Tags with the given tags. The provided slice is
+	// not used after the function returns, and may be re-used by the
+	// caller.
+	NewTags(src []string) *Tags
+
+	// NewUniqueTags creates a new *Tags. This method assumes the tags in the
+	// given slice are unique. The provided slice is not used after the
+	// function returns, and may be re-used by the caller.
+	NewUniqueTags(src ...string) *Tags
+
+	// NewTagsFromMap creates a new *Tags based on the keys of the given map.
+	NewTagsFromMap(src map[string]struct{}) *Tags
+
+	// NewTag creates a new *Tags with a single tag in it
+	NewTag(tag string) *Tags
+
+	// Builder constructors
+
+	// NewBuilder returns a fresh Builder
+	NewBuilder(capacity int) *Builder
+
+	// NewSliceBuilder returns a fresh Builder
+	NewSliceBuilder(levels, capacity int) *SliceBuilder
+
+	// Parsing
+
+	// ParseDSD parses a comma-separated set of tags, as used in the DogStatsD
+	// format.
+	ParseDSD(data []byte) (*Tags, error)
+
+	// Combination
+
+	// Union combines two *Tags instances that are not known to be
+	// disjoint. That is, there may exist tags that are in both tagsets.
+	Union(a, b *Tags) *Tags
+
+	// UnsafeDisjointUnion combines two disjoint *Tags instances.  The caller MUST
+	// ensure that the tagsets do not overlap. That is, there is no tag that
+	// is in both sets.  Failure to do so will result in undefined behavior.
+	//
+	// This function is only useful in rare circumstances, and its use should
+	// be accompanied by a comment explaining why the usage is safe.  In
+	// general, prefer to use Union.
+	UnsafeDisjointUnion(a, b *Tags) *Tags
```

Comment:
Is there a circumstance where we know we will need this? Otherwise I think we shouldn't include this at all.